File: 백준/Gold/2263. 트리의 순회/트리의 순회.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *

setrecursionlimit(10**9)
input = stdin.readline


def preOrder(inStart,inEnd,postStart,postEnd):
    
    global ans
    global in_order
    global post_order
    global node_pos
    
    if inStart > inEnd or postStart > postEnd:
        return

    root = post_order[postEnd] # 후위순회의 마지막은 항상 루트 노드
    pos = node_pos[root] # 중위순회 루트 인덱스
    
    
    #중위 순회 : 7 3 8 1 9 4 10  //(left) [0] 11 5 2 6
    #후위 순회 : 7 8 3 9 10 4 1 //(left) 11 5 6 2  [0]
    
    
    leftNode =  pos - inStart -1 # 왼쪽 서브트리 개수
    rightNode = inEnd - pos # 오른쪽 서브트리 개수
    
    # 이때 중위 순회에서 왼쪽 서브 트리의 범위는
    # 시작 인덱스 ~ (시작 인덱스 + 왼쪽 서브 트리의 노드 수)

    # 중위 순회에서 오른쪽 서브 트리의 범위는
    # (끝 인덱스 - 오른쪽 서브 트리의 수 + 1) ~ 끝 인덱스

    # 후위 순회에서 왼쪽 서브 트리의 범위는
    # 시작 인덱스 ~ (시작 인덱스 + 왼쪽 서브 트리의 노드 수)

    # 후위 순회에서 오른쪽 서브 트리의 범위는
    # (끝 인덱스 - 오른쪽 서브 트리의 수) ~ (끝 인덱스 - 1)
    
    ans.append(root) #전위 순회
    preOrder(inStart, inStart + leftNode, postStart, postStart + leftNode)
    preOrder(inEnd - rightNode + 1, inEnd, postEnd - rightNode, postEnd - 1)
    

n = int(input())

in_order = list(map(int,input().split()))
post_order = list(map(int,input().split()))
ans = []

node_pos = [-1 for _ in range(n+1)]

for i in range(len(in_order)):
    node_pos[in_order[i]] = i # 해당 노드에 대한 인덱스



preOrder(0,n-1,0,n-1)

print(*ans)


    




    

File: 백준/Gold/2263. 트리의 순회/트리의 순회.swift
import Foundation



    
let n = Int(readLine()!)!

var inOrder = readLine()!.split{ $0 == " " }.compactMap{ Int(String($0))! }
let postOrder = readLine()!.split{ $0 == " " }.compactMap{ Int(String($0))! }
var result:[String] = []

var node_pos:[Int] = [Int](repeating: -1, count: n+1)

for i in 0..<n {
    node_pos[inOrder[i]] = i // 해당 노드의 index
}


/*
           1
        2     3
     4   5  6    7
 

 inOrder =   4 2 5 1 6 3 7
 postOrder = 4 5 2 6 7 3 1

 */

func preOrder(inStart:Int,inEnd:Int,postStart:Int,postEnd:Int) {

    if inStart > inEnd || postStart > postEnd {
        return
    }

    let root = postOrder[postEnd]
    let rootIndex = node_pos[root]
    
    
    let leftNode = rootIndex - inStart - 1 // 왼쪽 노드 개수,   4 2 5
    let rightNode = inEnd - rootIndex // 오른쪽 노드 개수 , 6 3 7  ,   오른쪽 시작 인덱스
    
    //
    
    result.append("\(root)")
    preOrder(inStart: inStart, inEnd: inStart + leftNode, postStart: postStart, postEnd: postStart + leftNode)
    preOrder(inStart: inEnd - rightNode + 1, inEnd: inEnd, postStart: postEnd - rightNode, postEnd: postEnd - 1)  // postEnd - 1: 위에서 root로 사용했으므로 1개 빼고
}

preOrder(inStart: 0, inEnd: n-1, postStart: 0, postEnd: n-1)


print(result.joined(separator: " "))
    







